### PR TITLE
feat: Add support for resolving `Brillig` foreign calls

### LINKED ISSUE #27: Add support for resolving Brillig foreign calls
### Problem

[ACVM 0.13.0](https://github.com/noir-lang/acvm-simulator/pull/25) adds support for Brillig opcodes which the ACVM simulator should be able to process.

In particular we want to be able to process foreign calls in a similar way to how we handle oracle calls currently.

### Proposed solution

We should modify the existing `resolve_oracle` method to handle Brillig foreign calls.

### Alternatives considered

_No response_

### Additional context

_No response_

### Submission Checklist

- [X] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves #27 

## Summary\*

This PR adds support for resolving brillig foreign calls.

Draft until #43 is merged and we've added tests for this.

## Additional Context

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by TomAFrench** [comment]
Blocked by https://github.com/noir-lang/acvm/pull/390 as we need `ForeignCallWaitInfo` to be public.

File: src/foreign_calls.rs [hunk 1]
```diff
@@ -23,6 +28,36 @@ extern "C" {
...
+) -> Result<ForeignCallResult, String> {
+    // Prepare to call
+    let (name, inputs) = prepare_brillig_args(foreign_call_wait_info);
+
+    // Perform foreign call
+    let outputs = perform_foreign_call(foreign_call_callback, name, inputs).await?;
+    let outputs: Vec<Value> = vecmap(outputs, |output| output.into());
+    // The Brillig VM checks that the number of return values from
+    // the foreign call is valid so we don't need to do it here.
+    Ok(outputs.into())
```
Comment (TomAFrench) [inline src/foreign_calls.rs:43]: This makes the assumption that the foreign call will return a simple array of return values (`Vec<Value>`) whereas it returns a nested array (`Vec<Vec<Value>>`). I've been working on adding this support.

**Comment by TomAFrench** [comment]
This branch should not be merged until we properly support the brillig foreign call interface.

**Comment by TomAFrench** [comment]
I've added a testcase which covers the full range of the foreign call interface so this should be good to go.

**Review by sirasistant — APPROVED** [review]

**Pull request merged**